cpp/benchmarks/streaming/ndsh/join.cpp

```diff
@@ -344,6 +412,107 @@ streaming::Node inner_join_shuffle(
     co_await ch_out->drain(ctx->executor());
 }
 
+/**
+ * @brief Perform a left semi join between two tables, broadcasting the left table to all
+ * ranks.
+ *
+ * @param ctx Streaming context
+ * @param left Channel of `TableChunk`s used as the broadcasted build side. This table is
+ * broadcasted to all ranks.
+ * @param right Channel of `TableChunk`s joined in turn against the build side. This table
+ * is required to be shuffled / hash-partitioned.
+ * @param ch_out Output channel of `TableChunk`s.
+ * @param left_on Column indices of the keys in the left table.
+ * @param right_on Column indices of the keys in the right table.
+ * @param tag Disambiguating tag for the broadcast of the left table.
+ * @param keep_keys Does the result contain the key columns, or only "carrier" value
+ * columns
+ * @return Coroutine representing the completion of the join.
+ *
+ * @note This implementation assumes that:
+ * - `left` is small and fits in memory
+ * - `right` is shuffled / hash-partitioned
+ * It doesn't implement build table reuse across chunks of `left`, because we assume that
+ * `right` is too large to broadcast.
+ */
+streaming::Node left_semi_join_broadcast_left(
+    std::shared_ptr<streaming::Context> ctx,
+    std::shared_ptr<streaming::Channel> left,
+    std::shared_ptr<streaming::Channel> right,
+    std::shared_ptr<streaming::Channel> ch_out,
+    std::vector<cudf::size_type> left_on,
+    std::vector<cudf::size_type> right_on,
+    OpID tag,
+    [[maybe_unused]] KeepKeys keep_keys
+) {
+    streaming::ShutdownAtExit c{left, right, ch_out};
+    co_await ctx->executor()->schedule();
+    ctx->comm()->logger().print("Inner broadcast join ", static_cast<int>(tag));
+    auto left_table = to_device(
+        ctx, (co_await broadcast(ctx, left, tag)).release<streaming::TableChunk>()
+    );
+    ctx->comm()->logger().print(
+        "Left (probe) table has ", left_table.table_view().num_rows(), " rows"
+    );
+
+    std::size_t sequence = 0;
+    while (true) {
+        auto right_msg = co_await right->receive();
+        if (right_msg.empty()) {
+            break;
+        }
+        co_await ch_out->send(semi_join_chunk(
+            ctx,
+            left_table,
+            right_msg.release<streaming::TableChunk>(),
+            left_on,
+            right_on,
+            sequence++
+        ));
+    }
```

Comment:
Given that you're broadcasting the left table, I would do:

```
left = broadcast(...);
CudaEvent event;
event.record(left.stream());
semi_join_chunk(...., &event, left.stream());
```

and then `semi_join_chunk` can do:

```
chunk = co_await ch_right->receive();
chunk = to_device(...);
// ensure left table is ready
event.stream_wait(chunk.stream());
// do all join on chunk.stream()
...
// ensure deallocation of left happens after join
cuda_stream_join(left.stream(), chunk.stream(), ...)
```
WDYT?